src/util/spscfifo.h

```diff
@@ -0,0 +1,260 @@
+        DEBUG_ASSERT(write_count == count);
+    }
+
+    // Blocking write operation (copy) that loops until all provided data
+    // has been written
+    void push_all_copy(const value_type* data, size_type count = 1) {
+        size_type write_count = 0;
+        while (write_count < count) {
+            write_count += push(data + write_count, count - write_count);
+        }
+        DEBUG_ASSERT(write_count == count);
+    }
+
+    // Convenience method for writing a single element by value
+    bool try_push(value_type data) {
+        return push(&data, 1) == 1;
+    }
+
+    ///////////////////////////////////////////////////////////////////
+    // Consumer operations
+    ///////////////////////////////////////////////////////////////////
+
+    // Check how many elements are immediately available for reading.
+    // With min_count = 0 only relaxed memory ordering is applied
+    // when accessing atomic values.
+    size_type pop_peek(size_type min_count = 0) const {
+        return acquire_readable(min_count);
+    }
+
+    // Non-blocking skip read operation, i.e. only advance the
+    // read position
+    size_type pop_skip(size_type count = 1) {
+        auto readable = math_min(acquire_readable(count), count);
+        if (readable > 0) {
+            DEBUG_ASSERT(m_cback < m_size);
+            DEBUG_ASSERT(readable < m_size);
+            DEBUG_ASSERT(m_cback < m_cback + readable); // no overflow
+            m_cback = (m_cback + readable) % m_size;
+            DEBUG_ASSERT(m_cback < m_size);
+            m_back.store(m_cback, std::memory_order_release);
+        }
+        DEBUG_ASSERT(readable <= count);
+        return readable;
+    }
+
+    // Non-blocking read operation
+    size_type pop(value_type* data, size_type count = 1) {
+        DEBUG_ASSERT(data || !count);
+        auto readable = math_min(acquire_readable(count), count);
+        if (readable > 0) {
+            DEBUG_ASSERT(m_cback < m_size);
+            DEBUG_ASSERT(readable < m_size);
+            for (size_type i = 0; i < readable; ++i) {
+                data[i] = std::move(m_data[(m_cback + i) % m_size]);
+            }
+            DEBUG_ASSERT(m_cback < m_cback + readable); // no overflow
+            m_cback = (m_cback + readable) % m_size;
+            DEBUG_ASSERT(m_cback < m_size);
+            m_back.store(m_cback, std::memory_order_release);
+        }
+        DEBUG_ASSERT(readable <= count);
+        return readable;
+    }
+
+    // Blocking read operation that loops until all requested data
+    // has been read
+    void pop_all(value_type* data, size_type count = 1) {
+        size_type read_count = 0;
+        while (read_count < count) {
+            read_count += pop(data + read_count, count - read_count);
+        }
+        DEBUG_ASSERT(read_count == count);
+    }
+
+  private:
+    static size_type adjust_capacity(size_type capacity) {
+        VERIFY_OR_DEBUG_ASSERT(capacity <= std::numeric_limits<size_type>::max() / 2) {
```

Comment:
Why we limit the capacity to half the possible size?

The % trick only works if there is no rest after an overflow. We need to adjust the size for this.

Let's say size_type is char.
We have possible index 0...255
If capacity is 100 we have 100 + 1 (vor whatever reason)
Index 255 % 101 = 53 
Index (255 + 1) % 101 = 0

If we force the capacity to 128, it works
Index 255 % 128 = 127 (last index)
Index (255 + 1) % 127 = 0 (first index)

This is done here: https://github.com/mixxxdj/mixxx/blob/2944e2f2214e10e617e31b5f0c13d2938adf177f/src/util/fifo.h#L19